funsor/torch.py

```diff
@@ -234,6 +237,67 @@ def eager_reduce(self, op, reduced_vars):
             return Tensor(data, inputs, self.dtype)
         return super(Tensor, self).eager_reduce(op, reduced_vars)
 
+    def sample(self, sampled_vars, sample_inputs=None):
+        assert self.output == reals()
+        sampled_vars = sampled_vars.intersection(self.inputs)
+        if not sampled_vars:
+            return self
+
+        # Partition inputs into sample_inputs + batch_inputs + event_inputs.
+        if sample_inputs is None:
+            sample_inputs = OrderedDict()
+        assert frozenset(sample_inputs).isdisjoint(self.inputs)
+        sample_shape = tuple(int(d.dtype) for d in sample_inputs.values())
+        batch_inputs = OrderedDict((k, d) for k, d in self.inputs.items() if k not in sampled_vars)
+        event_inputs = OrderedDict((k, d) for k, d in self.inputs.items() if k in sampled_vars)
+        be_inputs = batch_inputs.copy()
+        be_inputs.update(event_inputs)
+        sb_inputs = sample_inputs.copy()
+        sb_inputs.update(batch_inputs)
+
+        # Sample all variables in a single Categorical call.
+        logits = align_tensor(be_inputs, self)
+        batch_shape = logits.shape[:len(batch_inputs)]
+        flat_logits = logits.reshape(batch_shape + (-1,))
+        sample_shape = tuple(d.dtype for d in sample_inputs.values())
+        flat_sample = torch.distributions.Categorical(logits=flat_logits).sample(sample_shape)
+        assert flat_sample.shape == sample_shape + batch_shape
+        results = []
+        mod_sample = flat_sample
+        for name, domain in reversed(list(event_inputs.items())):
+            size = domain.dtype
+            point = Tensor(mod_sample % size, sb_inputs, size)
+            mod_sample = mod_sample / size
+            results.append(Delta(name, point))
+
+        # Account for the log normalizer factor.
+        # Derivation: Let f be a nonnormalized distribution (a funsor), and
+        #   consider operations in linear space (source code is in log space).
+        #   Let x0 ~ f/|f| be a monte carlo sample from a normalized f/|f|.
+        #                              f(x0) / |f|      # dice numerator
+        #   Let g = delta(x=x0) |f| -----------------
+        #                           detach(f(x0)/|f|)   # dice denominator
+        #                       |detach(f)| f(x0)
+        #         = delta(x=x0) -----------------  be a dice approximation of f.
+        #                         detach(f(x0))
+        #   Then g is an unbiased estimator of f in value and all derivatives.
```

Comment:
Does this automatically do the maximum amount of gradient estimator Rao-Blackwellizing, like `TraceGraph_ELBO`?

Reply:
Hmm, I would expect Rao-Blackwellization to try to avoid calling `.sample()` at all and instead enumerate. Can you explain offline?